qmail.c

```diff
@@ -24,29 +24,39 @@ struct qmail *qq;
 {
   int pim[2];
   int pie[2];
+  int pierr[2];
 
   setup_qqargs();
 
   if (pipe(pim) == -1) return -1;
   if (pipe(pie) == -1) { close(pim[0]); close(pim[1]); return -1; }
+  if (pipe(pierr) == -1) {
+    close(pim[0]); close(pim[1]);
+    close(pie[0]); close(pie[1]);
+    return -1;
+  }
  
   switch(qq->pid = fork()) {
     case -1:
       close(pim[0]); close(pim[1]);
       close(pie[0]); close(pie[1]);
+      close(pierr[0]); close(pierr[1]);
       return -1;
     case 0:
       close(pim[1]);
       close(pie[1]);
+      close(pierr[0]); /* we want to receive data */
```

Comment:
We want to receive (read) data... and therefore we close the reading end (from where we read)... because now we are in the child and the child *sends* data?

If so, `/* we will send data to the parent */` would have gotten me faster to the point.  Maybe it is a fallacy of reading it as a patch and discovering it in-context makes sense though...